### PR TITLE
Liberty.bugfixes from.mitaka external gateway mode

### PR DESCRIPTION
**Pull request by ssorenso**
@jlongstaf 
#### What issues does this address?
This bug addresses several bugs.

#### What's this change do?
Provides the bugfixes as suggested in [this PR](https://github.com/F5Networks/f5-openstack-agent/pull/763) up to [this commit](https://github.com/F5Networks/f5-openstack-agent/pull/763/commits/45e6488088a01bdb0509a9fb96838c19003f0996).

#### Where should the reviewer start?
This is the cherrypick order:
* 6242965802b14146be3edc86d80242c44ac458d1
* a3dee1c77b303c93d58b84e741cd88771f1aea03
* 1f3c3f55f4c80c8222a244ed9baf1b8628176480
* 317a2bf5ba8ce564ff1dc5131bc96d410045b0d0
* b089f7645b1ac7bf42e4102a54b52343155dee0a
* 8f284a571340367283e58a9a4f84acdcc3185e6f
* 60a3ccb42439a9128843f351ca9997ed7d8aa09a
* 1cb674eab06fa1d70fb31af35f1c1925621c8b55
* 309382949d062615a35226cd8c1e093ae07f0e74
* 8d178d943e2af107686acba5b49fcc5b2bdc574c

These each come from the mitaka_external_gateway_mode branch.

#### Any background context?
See above for content source.  This is a refactor and port-back to liberty.

### CODE REVIEW

**Comment by ssorenso** [comment]
@jlongstaf, @richbrowne: I don't plan on making any more changes to this PR for now.  At least until it's reviewed and problems picked out.

**Comment by pjbreaux** [comment]
@ssorenso @jlongstaf:

Isolated one error with tearing down the loadbalancer. This is one failure we see in the smoke tests. I will report any others:

```
2017-09-28 10:51:38.266 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants [req-56f2ea9e-a6a8-4ec4-a10d-2d2bcab091f1 788d770bc2de44a1a729fb28857f6b05 35119d85090c4c02abf2bdf0063c80bd - - -] Folder deletion exception for tenant partition 35119d85090c4c02abf2bdf0063c80bd occurred. Manual cleanup might be required.
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants [req-56f2ea9e-a6a8-4ec4-a10d-2d2bcab091f1 788d770bc2de44a1a729fb28857f6b05 35119d85090c4c02abf2bdf0063c80bd - - -] 400 Unexpected Error: Bad Request for uri: https://10.190.7.249:443/mgmt/tm/sys/folder/~Project_35119d85090c4c02abf2bdf0063c80bd/
Text: u'{"code":400,"message":"0107082a:3: All objects must be removed from a partition (Project_35119d85090c4c02abf2bdf0063c80bd) before the partition may be removed, type ID (1846)","errorStack":[],"apiError":3}'
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants Traceback (most recent call last):
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants   File "/usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/tenants.py", line 120, in _remove_tenant_replication_mode
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants     self.system_helper.delete_folder(bigip, partition)
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants   File "/usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/system_helper.py", line 41, in delete_folder
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants     obj.delete()
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants   File "/usr/lib/python2.7/site-packages/f5/bigip/resource.py", line 1058, in delete
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants     self._delete(**kwargs)
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants   File "/usr/lib/python2.7/site-packages/f5/bigip/resource.py", line 1040, in _delete
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants     response = session.delete(delete_uri, **requests_params)
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants   File "/usr/lib/python2.7/site-packages/icontrol/session.py", line 272, in wrapper
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants     raise iControlUnexpectedHTTPError(error_message, response=response)
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants iControlUnexpectedHTTPError: 400 Unexpected Error: Bad Request for uri: https://10.190.7.249:443/mgmt/tm/sys/folder/~Project_35119d85090c4c02abf2bdf0063c80bd/
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants Text: u'{"code":400,"message":"0107082a:3: All objects must be removed from a partition (Project_35119d85090c4c02abf2bdf0063c80bd) before the partition may be removed, type ID (1846)","errorStack":[],"apiError":3}'
2017-09-28 10:51:38.267 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.tenants
```

**Comment by pjbreaux** [comment]
Another error that I see, possibly related to periodic resync:

```
2017-09-28 10:57:24.368 29336 DEBUG f5_openstack_agent.lbaasv2.drivers.bigip.plugin_rpc [-] f5_openstack_agent.lbaasv2.drivers.bigip.plugin_rpc.LBaaSv2PluginRPC method update_loadbalancer_status called with arguments (u'0c75a4f6-0928-40e5-9512-f14176eda0db', 'ACTIVE', 'ONLINE') {} wrapper /usr/lib/python2.7/site-packages/oslo_log/helpers.py:45
2017-09-28 10:57:24.371 29336 DEBUG f5_openstack_agent.lbaasv2.drivers.bigip.utils [-] sync request c9061d60-d11e-4b7e-a6e7-255e41433327 took 0.26370 secs wrapper /usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/utils.py:101
2017-09-28 10:57:24.371 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.agent_manager [-] Unable to retrieve ready service: 0c75a4f6-0928-40e5-9512-f14176eda0db
```

**Comment by pjbreaux** [comment]
@ssorenso @jlongstaf 

I also get a pool creation error. The pool is added to the database, but it is not associated properly to the virtual server:

```
0063c80bd - - -] Virtual server already exists create_listener /usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/listener_service.py:69
2017-09-28 10:59:57.883 29336 DEBUG root [req-d73752b5-36d9-48d1-8af7-3ff64a66dd89 788d770bc2de44a1a729fb28857f6b05 35119d85090c4c02abf2bdf0063c80bd - - -] post WITH uri: https://10.190.7.249:443/mgmt/tm/ltm/pool/ AND suffix:  AND kwargs: {'json': {'partition': u'Project_35119d85090c4c02abf2bdf0063c80bd', 'loadBalancingMode': 'round-robin', 'name': u'Project_886a2b63-d409-4801-b159-fcc1985e1355', 'description': u'Project_886a2b63-d409-4801-b159-fcc1985e1355:'}} wrapper /usr/lib/python2.7/site-packages/icontrol/session.py:257
2017-09-28 10:59:57.904 29336 DEBUG root [req-d73752b5-36d9-48d1-8af7-3ff64a66dd89 788d770bc2de44a1a729fb28857f6b05 35119d85090c4c02abf2bdf0063c80bd - - -] RESPONSE::STATUS: 200 Content-Type: application/json Content-Encoding: None
Text: u'{"kind":"tm:ltm:pool:poolstate","name":"Project_886a2b63-d409-4801-b159-fcc1985e1355","partition":"Project_35119d85090c4c02abf2bdf0063c80bd","fullPath":"/Project_35119d85090c4c02abf2bdf0063c80bd/Project_886a2b63-d409-4801-b159-fcc1985e1355","generation":92,"selfLink":"https://localhost/mgmt/tm/ltm/pool/~Project_35119d85090c4c02abf2bdf0063c80bd~Project_886a2b63-d409-4801-b159-fcc1985e1355?ver=12.1.2","allowNat":"yes","allowSnat":"yes","description":"Project_886a2b63-d409-4801-b159-fcc1985e1355:","ignorePersistedWeight":"disabled","ipTosToClient":"pass-through","ipTosToServer":"pass-through","linkQosToClient":"pass-through","linkQosToServer":"pass-through","loadBalancingMode":"round-robin","minActiveMembers":0,"minUpMembers":0,"minUpMembersAction":"failover","minUpMembersChecking":"disabled","queueDepthLimit":0,"queueOnConnectionLimit":"disabled","queueTimeLimit":0,"reselectTries":0,"serviceDownAction":"none","slowRampTime":10,"membersReference":{"link":"https://localhost/mgmt/tm/ltm/pool/~Project_35119d85090c4c02abf2bdf0063c80bd~Project_886a2b63-d409-4801-b159-fcc1985e1355/members?ver=12.1.2","isSubcollection":true}}' wrapper /usr/lib/python2.7/site-packages/icontrol/session.py:265
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver [req-d73752b5-36d9-48d1-8af7-3ff64a66dd89 788d770bc2de44a1a729fb28857f6b05 35119d85090c4c02abf2bdf0063c80bd - - -] listener
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver Traceback (most recent call last):
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver   File "/usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/icontrol_driver.py", line 1239, in _common_service_handler
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver     all_subnet_hints)
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver   File "/usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py", line 67, in assure_service
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver     self._assure_pools_configured(service)
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver   File "/usr/lib/python2.7/site-packages/f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py", line 233, in _assure_pools_configured
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver     raise f5_ex.PoolCreationException(err.message)
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver PoolCreationException: listener
2017-09-28 10:59:57.905 29336 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver
2017-09-28 10:59:57.907 29336 DEBUG f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver [req-d73752b5-36d9-48d1-8af7-3ff64a66dd89 788d770bc2de44a1a729fb28857f6b05 35119d85090c4c02abf2bdf0063c80bd - - -] f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver.iControlDriver method _update_pool_status called with arguments ([{u'lb_algorithm': u'ROUND_ROBIN', u'protocol': u'HTTP', u'description': u'', u'provisioning_status': 'ERROR', u'tenant_id': u'35119d85090c4c02abf2bdf0063c80bd', u'admin_state_up': True, u'session_persistence': None, u'healthmonitor_id': None, u'listeners': [{u'id': u'43763e32-ccab-4983-8349-3d4b9f6dc2fe'}], u'members': [], u'sessionpersistence': None, u'id': u'886a2b63-d409-4801-b159-fcc1985e1355', u'operating_status': u'OFFLINE', u'name': u'pool1'}],) {} wrapper /usr/lib/python2.7/site-packages/oslo_log/helpers.py:45
```

File: f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py [hunk 1]
```diff
@@ -155,7 +182,34 @@ def _assure_pools_created(self, service):
...
+        pools = service["pools"]
+        loadbalancer = service["loadbalancer"]
 
+        bigips = self.driver.get_config_bigips()
+
+        for pool in pools:
+            if pool['provisioning_status'] != plugin_const.PENDING_DELETE and \
+                    self._is_not_error(pool):
+                svc = {"loadbalancer": loadbalancer,
+                       "pool": pool}
```
Comment (jlongstaf) [inline f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py]: There's a problem with svc. It does not include a listener and it is later used to update the associated pool for a listener.

Comment (jlongstaf) [reply f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py]: Just need to add this:

 ` self.add_listener_pool(service, svc)`

**Comment by ssorenso** [comment]
@jlongstaf @pjbreaux @richbrowne: this should now work for smoke tests:
```shell
py.test --symbols ../symbols.json.example ./loadbalancer/ ./member/ ./listener/ ./pool
=========================================================================================== test session starts ============================================================================================
platform linux2 -- Python 2.7.6, pytest-3.2.2, py-1.4.34, pluggy-0.4.0
rootdir: /git/clones2/f5-openstack-agent, inifile:
plugins: symbols-0.1.0-a3, f5-sdk-2.3.3, cov-2.5.1
collected 22 items

loadbalancer/test_loadbalancer.py ..
loadbalancer/test_loadbalancer_admin_state.py .
loadbalancer/test_loadbalancer_disconnected.py ..
loadbalancer/test_loadbalancer_opflex.py ..
loadbalancer/test_loadbalancer_snat.py ...
loadbalancer/test_snat_common_network.py .
member/test_single_member.py ....
member/test_single_member_opflex.py .
member/test_single_member_weight_updates.py .
listener/test_listener_update.py .
listener/test_tcp_listener.py .
pool/test_pool_lb_method_change.py .
pool/test_session_persistence.py .
pool/test_single_pool.py .

======================================================================================= 22 passed in 246.92 seconds ========================================================================================
```

**Comment by ssorenso** [comment]
retest this please

File: f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py (hunk 1 above)
Comment (ssorenso) [reply f5_openstack_agent/lbaasv2/drivers/bigip/lbaas_builder.py]: @jlongstaf, @richbrowne This was handled earlier in this same method.  I don't know that I ever mentioned this.  So this should be fixed and the branch ready for review.

**Comment by ssorenso** [comment]
retest this please

**Comment by richbrowne** [comment]
retest this please

**Pull request merged**